primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -612,35 +634,35 @@ where
                 if first_index.is_none() {
                     first_index = Some(i);
                 }
-                if let LookupResult::Ok(_, proof) = self.inner.lookup(i) {
-                    proofs.push(proof);
-                } else {
-                    // The NMT is malformed, we cannot recover
-                    panic!()
-                }
+                proofs.push(self.lookup_proof(i));
             }
             let left_index = first_index.unwrap_or(0);
             let right_index = left_index + proofs.len() as u64;
             if left_index > 0 {
-                if let LookupResult::Ok(_, proof) = self.inner.lookup(left_index - 1) {
-                    left_boundary_proof = Some(proof);
-                } else {
-                    // The NMT is malformed, we cannot recover
-                    panic!()
-                }
+                left_boundary_proof = Some(self.lookup_proof(left_index - 1));
             }
             if right_index < self.num_leaves() - 1 {
-                if let LookupResult::Ok(_, proof) = self.inner.lookup(right_index + 1) {
-                    right_boundary_proof = Some(proof);
-                } else {
-                    // The NMT is malformed, we cannot recover
-                    panic!()
-                }
+                right_boundary_proof = Some(self.lookup_proof(right_index + 1));
             }
         } else {
-            proof_type = NamespaceProofType::Absence
-            // TODO: This only handles the simple absence proof case where the
-            // namespace is outside of the root's namespace range
+            proof_type = NamespaceProofType::Absence;
+            let ranges: Vec<&N> = self.namespace_ranges.keys().collect();
+            let closest_ns_idx = ranges
+                .binary_search_by(|leaf_ns| (**leaf_ns).cmp(&namespace))
+                .expect_err("The namespace should not be in the range");
```

Comment:
Instead of collecting a BST into a vec and then searching the vec, consider using [`range`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range), e.g. `self.namespace_ranges.range(namespace..).next()` will give you 
* the entry for `namespace` if `namespace` is in the map (should not be possible in this case)
* the first entry greater than `namespace` if there is one
* `None` if `namespace` is greater than all the keys in the map

Reply:
Suddenly I'm confused about the necessity of this `namesapce_ranges` as a `BTreeMap`. Isn't merkle tree itself already served as a similar data structure? Why should we have another BTreeMap.

Reply:
@mrain the `BTreeMap` just enables efficient lookup of namespace range, we could do without it but then we'd need to search the leaves linearly to generate namespace proofs.

Reply:
also @jbearer nice that looks cleaner, I'll try it out

Reply:
But the namespace part of the Merkle tree already forms a [segment tree](https://en.wikipedia.org/wiki/Segment_tree), right? It does support efficient lookup, with similar complexity of `BTreeMap`.

Reply:
I'm not sure I'm following, I guess with the leaves tagged as namespaces the Merkle tree implicitly represents a segment tree, but we still need to store the ranges somewhere right?

Reply:
Not really. You don't have to store the ranges, but you can easily implements the [`range`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) operation in time complexity `O(D)` where `D` is the tree height, usually in `log(N)` which is similar to `BTreeMap`.

Well, this can be an `improvement` since it reduces the storage overhead. We don't have to do it right now.

Reply:
I think the issue is that the ranges aren't actually stored in each internal node, they're computed dynamically from the leaves up when hashing. Is that right @nomaxg ?

Reply:
Oh gotcha - I see what you're saying. We might still be better off because the complexity of the BTreeMap is in the # of namespaces (which is strictly less than N). But yeah, if each namespace is sparse I guess we would be adding extra storage overhead for a similar lookup.

Reply:
Ahh @jbearer , IIUC the range information is stored in the `NamespacedHash`. There are 2 fields, `min_namespace` and `max_namespace`.